klippy/extras/ads1220.py

```diff
@@ -0,0 +1,203 @@
+        self.clock_updater.setup_query_command(
+            self.mcu, "query_ads1220_status oid=%c", self.oid, cq=cmdqueue)
+
+    def get_mcu(self):
+        return self.mcu
+
+    def get_samples_per_second(self):
+        return self.sps
+
+    # returns a tuple of the minimum and maximum value of the sensor, used to
+    # detect if a data value is saturated
+    def get_range(self):
+        range_max = (2 ** 24)
+        return -range_max, range_max
+
+    # add_client interface, direct pass through to bulk_sensor API
+    def add_client(self, callback):
+        self.batch_bulk.add_client(callback)
+
+    # Measurement decoding
+    def _extract_samples(self, raw_blocks):
+        # local variables to optimize inner loop below
+        unpack_int32 = self._unpack_int32
+        # Process every message in capture_buffer
+        max_samples = (len(raw_blocks) * MAX_SAMPLES_PER_MESSAGE)
+        samples = collections.deque(maxlen=max_samples)
+        timestamps = TimestampHelper(self.clock_sync, self.clock_updater,
+                                     MAX_SAMPLES_PER_MESSAGE)
+        for block in raw_blocks:
+            timestamps.update_sequence(block['sequence'])
+            data = bytearray(block['data'])
+            for i in range(len(data) // BYTES_PER_SAMPLE):
+                counts = unpack_int32(data, offset=BYTES_PER_SAMPLE * i)[0]
+                samples.append((timestamps.time_of_msg(i), counts))
+        timestamps.set_last_chip_clock()
+        return list(samples)
+
+    # Start, stop, and process message batches
+    def _start_measurements(self):
+        # Start bulk reading
+        self.bulk_queue.clear_samples()
+        self.reset_chip()
+        self.setup_chip()
+        rest_ticks = self.mcu.seconds_to_clock(0.7 / float(self.sps))
+        self.query_ads1220_cmd.send([self.oid, rest_ticks])
+        logging.info("ADS1220 starting '%s' measurements", self.name)
+        # Initialize clock tracking
+        self.clock_updater.note_start()
+
+    def _finish_measurements(self):
+        # Halt bulk reading
+        # TODO: this logs errors when klipper shuts down for other reasons
+        self.query_ads1220_cmd.send_wait_ack([self.oid, 0])
+        self.bulk_queue.clear_samples()
+        logging.info("ADS1220 finished '%s' measurements", self.name)
+
+    def _process_batch(self, eventtime):
+        self.clock_updater.update_clock()
+        raw_samples = self.bulk_queue.pull_samples()
+        if not raw_samples:
+            return {}
+        samples = self._extract_samples(raw_samples)
+        if not samples:
+            return {}
+        return {'data': samples}
+
+    def reset_chip(self):
+        # the reset command takes 50us to complete
+        self.send_command(RESET_CMD)
+        # TODO: is a manual 50us delay needed here? Testing says it's not.
+        # read startup register state and validate
+        val = self.read_reg(0x0, 4)
+        if hexify(val) != hexify(RESET_STATE):
```

Comment:
FWIW, it's certainly seems odd to encode two numeric objects into a string to compare them..